src/ir/auth_logic/ast.h

```diff
@@ -139,6 +139,55 @@ class Assertion {
   AssertionVariantType value_;
 };
 
+enum class ConstantType { kNumberType, kPrincipalType };
+
+// Relation declaration types are of 3 forms
+// 1. NumberType
+// 2. PrincipalType
+// 3. User defined CustomType (string to store name of the type)
+class ArgumentType {
+ public:
+  enum class Kind { kNumber, kPrincipal, kCustom };
+  explicit ArgumentType(Kind kind, std::string name)
+      : kind_(std::move(kind)), name_(std::move(name)) {}
+  const Kind& kind() const { return kind_; }
+  const std::string name() const { return name_; }
+
+ private:
+  Kind kind_;
+  std::string name_;
+};
+
+class Argument {
+ public:
+  explicit Argument(std::string argument_name, ArgumentType argument_type)
+      : argument_name_(std::move(argument_name)),
+        argument_type_(std::move(argument_type)) {}
+  const std::string& argument_name() { return argument_name_; }
```

Comment:
Add const declarations at the end, which marks the members as const functions. Also, return `ArgumentType` instead of `const ArgumentType`
```
  const std::string& argument_name() const  { return argument_name_; } 
  ArgumentType argument_type() const { return argument_type_; }
```